File: app/src/main/java/com/tobiasfried/rasa/BrewsFragment.java
package com.tobiasfried.rasa;

import android.app.Application;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.Toast;

import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;

import com.tobiasfried.rasa.model.Brew;
import com.tobiasfried.rasa.viewmodel.MainViewModel;
import com.tobiasfried.rasa.viewmodel.MainViewModelFactory;

import java.util.Objects;

import static com.tobiasfried.rasa.EntryActivity.EXTRA_BREW_ID;

public class BrewsFragment extends Fragment {

    private static final String LOG_TAG = BrewsFragment.class.getSimpleName();

    private MainViewModel viewModel;
    private FirebaseFirestore db;
    private FirestoreRecyclerAdapter<Brew, BrewViewHolder> mAdapter;
    private FirestoreRecyclerOptions<Brew> options;
    private Query query;
    private String sortOptions;
    private Query.Direction sortOrder = Query.Direction.ASCENDING;
    private Brew deleted;
    private Unbinder unbinder;

    private View rootView;

    @BindView(R.id.list)
    RecyclerView recyclerView;

    @BindView(R.id.spinner_sort_by)
    Spinner sortSpinner;

    @BindView(R.id.button_sort_order)
    MaterialButton sortOrderButton;

    public BrewsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Get Database instance
        db = FirebaseFirestore.getInstance();
        MainViewModelFactory factory = new MainViewModelFactory(db);
        viewModel = ViewModelProviders.of(this, factory).get(MainViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_brews, container, false);
        unbinder = ButterKnife.bind(this, rootView);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));

        // Set Spinner Adapter
        ArrayAdapter<CharSequence> sortAdapter = ArrayAdapter.createFromResource(Objects.requireNonNull(getContext()),
                R.array.array_sort_names, R.layout.spinner_item_sort);
        sortSpinner.setAdapter(sortAdapter);
        sortSpinner.setSelection(0);
        sortSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                sortOptions = getResources().getStringArray(R.array.array_sort_options)[position];
                setupRecyclerView();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        // Set Sort Order Button
        sortOrderButton.setOnClickListener(v -> {
            float rotation = v.getRotation();
            v.setRotation(rotation == 0 ? 180 : 0);
            sortOrder = rotation == 0 ? Query.Direction.ASCENDING : Query.Direction.DESCENDING;
            setupRecyclerView();
        });

        return rootView;
    }

    private void setupRecyclerView() {
        // Inflate and setup RecyclerView
        query = db.collection(Brew.CURRENT)
                .orderBy(sortOptions, sortOrder);

        options = new FirestoreRecyclerOptions.Builder<Brew>()
                .setQuery(query, Brew.class)
                .setLifecycleOwner(this)
                .build();

        mAdapter = new FirestoreRecyclerAdapter<Brew, BrewViewHolder>(options) {

            private int expandedItemIndex = -1;

            @NonNull
            @Override
            public BrewViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                View itemView = getLayoutInflater().inflate(R.layout.list_item_brew, parent, false);
                return new BrewViewHolder(itemView);
            }

            @Override
            protected void onBindViewHolder(@NonNull BrewViewHolder holder, int position, @NonNull Brew brew) {
                // Bind Views
                holder.bind(brew);

                // Set expander ClickListener
                holder.card.setOnClickListener(v -> {
                    if (holder.getAdapterPosition() == expandedItemIndex) {
                        notifyItemChanged(holder.getAdapterPosition());
                        expandedItemIndex = -1;
                    } else {
                        if (expandedItemIndex != -1) {
                            notifyItemChanged(expandedItemIndex);
                        }
                        expandedItemIndex = holder.getAdapterPosition();
                        notifyItemChanged(holder.getAdapterPosition());
                    }
                });

                if (position == expandedItemIndex) {
                    holder.quickActions.setVisibility(View.VISIBLE);
                } else {
                    holder.quickActions.setVisibility(View.GONE);
                }

                // Set quick action ClickListeners
                holder.details.setOnClickListener(v -> {
                    String brewId = getSnapshots().getSnapshot(holder.getAdapterPosition()).getId();
                    Intent intent = new Intent(getActivity(), EntryExtendedActivity.class);
                    intent.putExtra(EXTRA_BREW_ID, brewId);
                    startActivity(intent);
                    holder.expanded = false;
                    //notifyItemRangeChanged(0, mAdapter.getItemCount());
                });

                holder.markComplete.setOnClickListener(v -> {
                    String brewId = mAdapter.getSnapshots().getSnapshot(holder.getAdapterPosition()).getId();
                    if (!brew.advanceStage()) {
                        brew.getSecondaryFerment().second = System.currentTimeMillis();
                        db.collection(Brew.HISTORY).add(brew);
                        deleted = mAdapter.getSnapshots().getSnapshot(holder.getAdapterPosition()).toObject(Brew.class);
                        db.collection(Brew.CURRENT).document(brewId).delete();

                        // Show Snackbar with undo action
                        Snackbar.make(rootView, "Brew marked complete", Snackbar.LENGTH_LONG)
                                .setAction("UNDO", v2 -> {
                                    db.collection(Brew.CURRENT).add(deleted).addOnCompleteListener(task -> {
                                        if (task.isSuccessful()) {
                                            deleted = null;
                                        }
                                    });
                                    // TODO get the correct history collection ID
                                    db.collection(Brew.HISTORY).document(brewId).delete();
                                })
                                .setActionTextColor(getResources().getColor(android.R.color.white, Objects.requireNonNull(getActivity()).getTheme()))
                                .show();
                    } else {
                        db.collection(Brew.CURRENT).document(brewId).set(brew);
                    }
                    holder.expanded = false;
                    notifyItemChanged(holder.getAdapterPosition());
                });

                holder.delete.setOnClickListener(v -> {
                    String brewId = mAdapter.getSnapshots().getSnapshot(holder.getAdapterPosition()).getId();
                    deleted = mAdapter.getSnapshots().getSnapshot(holder.getAdapterPosition()).toObject(Brew.class);
                    db.collection(Brew.CURRENT).document(brewId).delete();

                    // Show Snackbar with undo action
                    Snackbar.make(rootView, "Brew Deleted", Snackbar.LENGTH_LONG)
                            .setAction("UNDO", h -> db.collection(Brew.CURRENT).add(deleted).addOnCompleteListener(task -> {
                                if (task.isSuccessful()) {
                                    deleted = null;
                                }
                            }))
                            .setActionTextColor(getResources().getColor(android.R.color.white, Objects.requireNonNull(getActivity()).getTheme()))
                            .show();

                    holder.expanded = false;
                    expandedItemIndex = -1;
                    notifyItemRangeChanged(0, mAdapter.getItemCount());
                });

            }

            @Override
            public void onViewRecycled(@NonNull BrewViewHolder holder) {
                super.onViewRecycled(holder);
                holder.expanded = false;
            }

        };

        // Attach Touch Listener
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                if (direction == ItemTouchHelper.RIGHT) {
                    // Move brew to recently deleted and delete it from database
                    String brewId = mAdapter.getSnapshots().getSnapshot(viewHolder.getAdapterPosition()).getId();
                    deleted = mAdapter.getSnapshots().getSnapshot(viewHolder.getAdapterPosition()).toObject(Brew.class);
                    // TODO delete view MainViewModel
                    db.collection(Brew.CURRENT).document(brewId).delete();

                    // Show Snackbar with undo action
                    Snackbar.make(rootView, "Brew Deleted", Snackbar.LENGTH_LONG)
                            .setAction("UNDO", v -> db.collection(Brew.CURRENT).add(deleted).addOnCompleteListener(task -> {
                                if (task.isSuccessful()) {
                                    deleted = null;
                                }
                            }))
                            .setActionTextColor(getResources().getColor(android.R.color.white, Objects.requireNonNull(getActivity()).getTheme()))
                            .show();
                }
            }
        }).attachToRecyclerView(recyclerView);
        ((SimpleItemAnimator) Objects.requireNonNull(recyclerView.getItemAnimator())).setSupportsChangeAnimations(false);

        recyclerView.setAdapter(mAdapter);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
}

File: app/src/main/java/com/tobiasfried/rasa/brews/BrewsFragment.kt
package com.tobiasfried.rasa.brews

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SimpleItemAnimator
import butterknife.BindView
import butterknife.Unbinder

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner

import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.google.android.material.button.MaterialButton
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query

import com.tobiasfried.rasa.domain.Brew
import com.tobiasfried.rasa.MainViewModel
import com.tobiasfried.rasa.MainViewModelFactory

import java.util.Objects

import com.tobiasfried.rasa.entry.EntryExtendedActivity
import com.tobiasfried.rasa.R
import com.tobiasfried.rasa.databinding.FragmentBrewsBinding

class BrewsFragment : Fragment() {

    private lateinit var db: FirebaseFirestore
    private var mAdapter: FirestoreRecyclerAdapter<Brew, BrewViewHolder>? = null
    private var options: FirestoreRecyclerOptions<Brew>? = null
    private var query: Query? = null
    private var sortOptions: String? = null
    private var sortOrder: Query.Direction = Query.Direction.ASCENDING
    private var deleted: Brew? = null
    private var unbinder: Unbinder? = null

    private var rootView: View? = null

    @BindView(R.id.button_sort_order)
    internal var sortOrderButton: MaterialButton? = null

    private lateinit var binding: FragmentBrewsBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Get Database instance
        db = FirebaseFirestore.getInstance()

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentBrewsBinding.inflate(inflater)

        // Allows Data Binding to Observe LiveData with the lifecycle of this Fragment
        binding.lifecycleOwner = this

        val factory = MainViewModelFactory(db)
        binding.viewModel = ViewModelProviders.of(this, factory).get(BrewsViewModel::class.java)

        binding.list.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)

        // Set Spinner Adapter
        val sortAdapter = ArrayAdapter.createFromResource(context,
                R.array.array_sort_names, R.layout.spinner_item_sort)
        binding.spinnerSortBy.adapter = sortAdapter
        binding.spinnerSortBy.setSelection(0)
        binding.spinnerSortBy.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                sortOptions = resources.getStringArray(R.array.array_sort_options)[position]
                setupRecyclerView()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        // Set Sort Order Button
        sortOrderButton!!.setOnClickListener { v ->
            val rotation = v.rotation
            v.rotation = (if (rotation == 0f) 180 else 0).toFloat()
            sortOrder = if (rotation == 0f) Query.Direction.ASCENDING else Query.Direction.DESCENDING
            setupRecyclerView()
        }

        return rootView
    }

    private fun setupRecyclerView() {
        // Inflate and setup RecyclerView
        query = db!!.collection(Brew.CURRENT)
                .orderBy(sortOptions!!, sortOrder)

        options = FirestoreRecyclerOptions.Builder<Brew>()
                .setQuery(query!!, Brew::class.java)
                .setLifecycleOwner(this)
                .build()

        mAdapter = object : FirestoreRecyclerAdapter<Brew, BrewViewHolder>(options!!) {

            private var expandedItemIndex = -1

            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BrewViewHolder {
                val itemView = layoutInflater.inflate(R.layout.list_item_brew, parent, false)
                return BrewViewHolder(itemView)
            }

            override fun onBindViewHolder(holder: BrewViewHolder, position: Int, brew: Brew) {
                // Bind Views
                holder.bind(brew)

                // Set expander ClickListener
                holder.card!!.setOnClickListener { v ->
                    if (holder.adapterPosition == expandedItemIndex) {
                        notifyItemChanged(holder.adapterPosition)
                        expandedItemIndex = -1
                    } else {
                        if (expandedItemIndex != -1) {
                            notifyItemChanged(expandedItemIndex)
                        }
                        expandedItemIndex = holder.adapterPosition
                        notifyItemChanged(holder.adapterPosition)
                    }
                }

                if (position == expandedItemIndex) {
                    holder.quickActions!!.visibility = View.VISIBLE
                } else {
                    holder.quickActions!!.visibility = View.GONE
                }

                // Set quick action ClickListeners
                holder.details!!.setOnClickListener { v ->
                    val brewId = snapshots.getSnapshot(holder.adapterPosition).id
                    val intent = Intent(activity, EntryExtendedActivity::class.java)
                    intent.putExtra(EXTRA_BREW_ID, brewId)
                    startActivity(intent)
                    holder.expanded = false
                    //notifyItemRangeChanged(0, mAdapter.getItemCount());
                }

                holder.markComplete!!.setOnClickListener { v ->
                    val brewId = mAdapter!!.snapshots.getSnapshot(holder.adapterPosition).id
                    if (!brew.advanceStage()) {
                        brew.secondaryFerment!!.second = System.currentTimeMillis()
                        db.collection(Brew.HISTORY).add(brew)
                        deleted = mAdapter!!.snapshots.getSnapshot(holder.adapterPosition).toObject(Brew::class.java)
                        db.collection(Brew.CURRENT).document(brewId).delete()

                        // Show Snackbar with undo action
                        Snackbar.make(binding.root, "Brew marked complete", Snackbar.LENGTH_LONG)
                                .setAction("UNDO") {
                                    db.collection(Brew.CURRENT).add(deleted!!).addOnCompleteListener { task ->
                                        if (task.isSuccessful) {
                                            deleted = null
                                        }
                                    }
                                    // TODO get the correct history collection ID
                                    db.collection(Brew.HISTORY).document(brewId).delete()
                                }
                                .setActionTextColor(resources.getColor(android.R.color.white, Objects.requireNonNull<FragmentActivity>(activity).getTheme()))
                                .show()
                    } else {
                        db!!.collection(Brew.CURRENT).document(brewId).set(brew)
                    }
                    holder.expanded = false
                    notifyItemChanged(holder.adapterPosition)
                }

                holder.delete!!.setOnClickListener { v ->
                    val brewId = mAdapter!!.snapshots.getSnapshot(holder.adapterPosition).id
                    deleted = mAdapter!!.snapshots.getSnapshot(holder.adapterPosition).toObject(Brew::class.java)
                    db!!.collection(Brew.CURRENT).document(brewId).delete()

                    // Show Snackbar with undo action
                    Snackbar.make(rootView!!, "Brew Deleted", Snackbar.LENGTH_LONG)
                            .setAction("UNDO") { h ->
                                db!!.collection(Brew.CURRENT).add(deleted!!).addOnCompleteListener { task ->
                                    if (task.isSuccessful) {
                                        deleted = null
                                    }
                                }
                            }
                            .setActionTextColor(resources.getColor(android.R.color.white, Objects.requireNonNull<FragmentActivity>(activity).getTheme()))
                            .show()

                    holder.expanded = false
                    expandedItemIndex = -1
                    notifyItemRangeChanged(0, mAdapter!!.itemCount)
                }

            }

            override fun onViewRecycled(holder: BrewViewHolder) {
                super.onViewRecycled(holder)
                holder.expanded = false
            }

        }

        // Attach Touch Listener
        ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                if (direction == ItemTouchHelper.RIGHT) {
                    // Move brew to recently deleted and delete it from database
                    val brewId = mAdapter!!.snapshots.getSnapshot(viewHolder.adapterPosition).id
                    deleted = mAdapter!!.snapshots.getSnapshot(viewHolder.adapterPosition).toObject(Brew::class.java)
                    // TODO delete view MainViewModel
                    db!!.collection(Brew.CURRENT).document(brewId).delete()

                    // Show Snackbar with undo action
                    Snackbar.make(rootView!!, "Brew Deleted", Snackbar.LENGTH_LONG)
                            .setAction("UNDO") { v ->
                                db!!.collection(Brew.CURRENT).add(deleted!!).addOnCompleteListener { task ->
                                    if (task.isSuccessful) {
                                        deleted = null
                                    }
                                }
                            }
                            .setActionTextColor(resources.getColor(android.R.color.white, Objects.requireNonNull<FragmentActivity>(activity).getTheme()))
                            .show()
                }
            }
        }).attachToRecyclerView(recyclerView)
        (Objects.requireNonNull<ItemAnimator>(recyclerView!!.itemAnimator) as SimpleItemAnimator).supportsChangeAnimations = false

        recyclerView!!.adapter = mAdapter
    }

    override fun onDestroyView() {
        super.onDestroyView()
        unbinder!!.unbind()
    }

    companion object {

        private val LOG_TAG = BrewsFragment::class.java.simpleName
    }
}// Required empty public constructor